apps/docs/docs/message-db/source.md

```diff
@@ -0,0 +1,61 @@
+# Source
+
+The MessageDB source continuously fetches batches of messages and dispatches
+them to a supplied handler and dispatches them to a supplied handler. It
+supports reading from multiple categories at a time, server side filtering, and
+splitting processing into multiple parallel consumer groups.
+
+## Options
+
+When creating a MessageDB source via `MessageDbSource.create` the following
+options can be supplied
+
+```ts
+interface CreateOptions {
+  /** The database pool to use to read messages from the category */
+  pool: Pool
+  /** The categories to read from */
+  categories: string[]
+  /** The maximum number of messages to read from the category at a time
+   * @default 500
+   */
+  batchSize?: number
+  /** The name of the consumer group to use for checkpointing */
+  groupName: string
+  /** The checkpointer to use for checkpointing */
+  checkpointer: ICheckpointer
+  /** The handler to call for each batch of stream messages */
+  handler: (streamName: string, events: ITimelineEvent<string>[]) => Promise<void>
+  /** The number of milliseconds to sleep between tail reads */
+  tailSleepIntervalMs: number
+  /** The maximum number of concurrent streams to process, enforced via p-limit */
+  maxConcurrentStreams: number
+  /** Apply a server side filter condition to the reading of messages
+   * NOTE: you will need to set the `message_store.sql_condition` setting to `"on"` to use this feature
+   */
+  condition?: string
+  /** When using consumer groups: the index of the consumer. 0 <= i <= consumerGroupSize
+   * each consumer in the group maintains their own checkpoint */
+  consumerGroupMember?: number
+  /** The number of group consumers you have deployed */
+  consumerGroupSize?: number
+}
+type Options = Omit<CreateOptions, "pool">
+
+```
+
+## Consumer groups
+
+When using the consumer group functionality two things deviate from normal
+operation.
+
+1. The checkpointer will checkpoint to `{groupName}-{member}`
+2. MessageDB will filter messages based on a hash of the stream id
+
+This means that each consumer in the group maintains a distinct checkpoint and
+is therefore not easily autoscaled. In order to expand or contract a consumer
+group you will need to wipe the checkpoints and have all of the consumers in the
+group start from the current lowest checkpoint.
+
+Using a hash of the ID guarantees that all messages for a particular entity will
+be handled by the same consumer.
```

Comment:
be handled in the correct order by the same consumer instance

(I think using stream instead of entity here will help convey what they need to be thinking)

Reply:
Message db uses the "cardinal id" to do its sharding, Cardinal ID being based on the eventide concept where a stream name is formed as `{category}-{cardinal_id}+{other_ids}` so we kind of opt out of the behaviour by using `_` as our stream id separator. But in the case where you have e.g. an `Invoice-1` and an `InvoiceAutoEmailer-1` you would be guaranteed that they're both handled by the same consumer instance and therefore would benefit from the caching layer most likely having an up-to-date view. Or at least that's how I imagine Scott's reasoning for the behaviour goes.

That said, maybe it's not worth calling out here

Reply:
Interesting. Can't think of a thing I've done or wanted to do which would benefit enough from this to warrant attempting to convey the concept in a terse form here. Ultimately if your dispatch is at stream level and concurrent, the concept is not relevant here. But you do want to convey that all messages will consistently route to the same consumer instance and hence guarantee no concurrent processing for a given stream.
Either way, overindexing on the term entity without layout out the concept in this context is not something to overindex on. While I'm sure it will speak to an audience, I still feel that entity is equally likely to get people thinking CRUD or "sending messages: rather than getting them in the right frame of mind for handling coalesced groups of events at positions in a sequential stream (I see lots of cases where people see the goal of an event contract/structure to be to double as a message that can be handled in one shot by a consumer. I'm all for using direct approaches that solve problems over tonnes of layers that some dogmatic best practice scheme suggests, but you still need to default to talking in terms that are helpful to conveying a good general strategy).

Reply:
agreed, I'll remove the reference to entity, I think it'll cause more harm than good